Data/ByteString/Short/Internal.hs

```diff
@@ -446,6 +668,899 @@ concat sbss =
+            w' = mask' .&. ((w `shiftL` 8) .|. b)
+    {-# INLINE shift #-}
+
+
+-- --------------------------------------------------------------------
+-- Searching ShortByteString
+
+-- | /O(n)/ 'elem' is the 'ShortByteString' membership predicate.
+--
+-- @since 0.11.3.0
+elem :: Word8 -> ShortByteString -> Bool
+elem c = \sbs -> case elemIndex c sbs of Nothing -> False ; _ -> True
+
+-- | /O(n)/ 'filter', applied to a predicate and a ByteString,
+-- returns a ByteString containing those characters that satisfy the
+-- predicate.
+--
+-- @since 0.11.3.0
+filter :: (Word8 -> Bool) -> ShortByteString -> ShortByteString
+filter k = \sbs -> let l = length sbs
+                   in if | l <= 0    -> sbs
+                         | otherwise -> createAndTrim' l $ \mba -> go mba (asBA sbs) l
+  where
+    go :: forall s. MBA s -- mutable output bytestring
+       -> BA              -- input bytestring
+       -> Int             -- length of input bytestring
+       -> ST s Int
+    go !mba ba !l = go' 0 0
+      where
+        go' :: Int -- bytes read
+            -> Int -- bytes written
+            -> ST s Int
+        go' !br !bw
+          | br >= l   = return bw
+          | otherwise = do
+              let w = indexWord8Array ba br
+              if k w
+              then do
+                writeWord8Array mba bw w
+                go' (br+1) (bw+1)
+              else
+                go' (br+1) bw
+
+-- | /O(n)/ The 'find' function takes a predicate and a ByteString,
+-- and returns the first element in matching the predicate, or 'Nothing'
+-- if there is no such element.
+--
+-- > find f p = case findIndex f p of Just n -> Just (p ! n) ; _ -> Nothing
+--
+-- @since 0.11.3.0
+find :: (Word8 -> Bool) -> ShortByteString -> Maybe Word8
+find f = \sbs -> case findIndex f sbs of
+                    Just n -> Just (sbs `index` n)
+                    _      -> Nothing
+
+-- | /O(n)/ The 'partition' function takes a predicate a ByteString and returns
+-- the pair of ByteStrings with elements which do and do not satisfy the
+-- predicate, respectively; i.e.,
+--
+-- > partition p bs == (filter p sbs, filter (not . p) sbs)
+--
+-- @since 0.11.3.0
+partition :: (Word8 -> Bool) -> ShortByteString -> (ShortByteString, ShortByteString)
+partition k = \sbs -> let l = length sbs
+                   in if | l <= 0    -> (sbs, sbs)
+                         | otherwise -> createAndTrim'' l $ \mba1 mba2 -> go mba1 mba2 (asBA sbs) l
+  where
+    go :: forall s.
+          MBA s           -- mutable output bytestring1
+       -> MBA s           -- mutable output bytestring2
+       -> BA              -- input bytestring
+       -> Int             -- length of input bytestring
+       -> ST s (Int, Int)
+    go !mba1 !mba2 ba !l = go' 0 0 0
+      where
+        go' :: Int -- bytes read
+            -> Int -- bytes written to b1
+            -> Int -- bytes written to b2
+            -> ST s (Int, Int)
+        go' !br !bw1 !bw2
```

Comment:
One of these counters can be computed from the other two, e.g. `bw2 = br - bw1`. How does this affect the benchmarks?

Reply:
I can't do benchmarking reliably on my thinkpad. It's full of CPU throttling and whatnot.

Reply:
Alright. I have added this task to https://github.com/haskell/bytestring/issues/350.

Reply:
Orig:
```
        mostlyTrueFast:  OK (0.35s)
          78.3 μs ± 5.0 μs
        mostlyFalseFast: OK (0.28s)
          63.8 μs ± 4.9 μs
        balancedFast:    OK (0.25s)
          119  μs ± 6.8 μs
        mostlyTrueSlow:  OK (0.28s)
          2.18 ms ± 104 μs
        mostlyFalseSlow: OK (0.14s)
          2.14 ms ± 193 μs
        balancedSlow:    OK (0.14s)
          2.21 ms ± 183 μs
```

With optimization: 
```
        mostlyTrueFast:  OK (0.36s)
          73.7 μs ± 5.4 μs
        mostlyFalseFast: OK (0.26s)
          61.6 μs ± 2.9 μs
        balancedFast:    OK (0.23s)
          110  μs ± 8.7 μs
        mostlyTrueSlow:  OK (0.14s)
          2.16 ms ± 181 μs
        mostlyFalseSlow: OK (0.14s)
          2.15 ms ± 192 μs
        balancedSlow:    OK (0.14s)
          2.20 ms ± 206 μs
```

I think the only one that's noticable is `balancedFast` and `mostlyTrueFast` and that seems to be somewhat consistent across many reruns.